core/shared/src/main/scala/zio/ZIOConcurrent.scala

```diff
@@ -0,0 +1,153 @@
+package zio
+
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.ConcurrentLinkedQueue
+import scala.annotation.tailrec
+import zio.Synchronizer._
+
+final class CountDownLatch(count: Int, maxWaiters: Option[Int]) {
+  private[this] val sync = Synchronizer(count, maxWaiters)
+
+  // TODO is it possible to evaluate once and reuse the Promise?
+  private[this] val awaitZero = sync.acquire(count)
+
+  def await: UIO[Unit] = awaitZero.asInstanceOf[UIO[Unit]]
+
+  // TODO stop at zero
+  def countDown(): UIO[Unit] = sync.release(1)
+}
+
+final class AsyncSemaphore(initialPermits: Int, maxWaiters: Option[Int]) {
+  private[this] val sync = Synchronizer(initialPermits, maxWaiters)
+
+  def apply[R, E, A](zio: ZIO[R, E, A]): ZIO[R, Either[E, Full], A] =
+    sync
+      .acquire(1)
+      .mapError(Right(_))
+      .flatMap { _ =>
+        zio
+          .mapError(Left(_))
+          .ensuring(sync.release(1))
```

Comment:
is it possible that the computation is cancelled after `acquire` but before it reaches this point?

Reply:
Yes interruption can in general be observed between every continuation. You can prevent interruption from being observed in a particular region by using `uninterruptibleMask` though that does have some performance cost. If you do a bunch of things within a single effect like `effectSuspendTotal` below interruption cannot be observed within that so that can be another alternative.